contracts/crafting/facets/CraftingFacet.sol

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * Authors: Moonstream Engineering (engineering@moonstream.to)
+ * GitHub: https://github.com/bugout-dev/dao
+ */
+
+pragma solidity ^0.8.0;
+import "@openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import "@openzeppelin-contracts/contracts/token/ERC20/presets/ERC20PresetMinterPauser.sol";
+import "@openzeppelin-contracts/contracts/token/ERC1155/extensions/ERC1155Burnable.sol";
+import "@openzeppelin-contracts/contracts/token/ERC721/extensions/ERC721Burnable.sol";
+import "@openzeppelin-contracts/contracts/token/ERC1155/utils/ERC1155Holder.sol";
+import "@openzeppelin-contracts/contracts/token/ERC721/utils/ERC721Holder.sol";
+import {MockTerminus} from "../../mock/MockTerminus.sol";
+import {MockErc20} from "../../mock/MockErc20.sol";
+import "../libraries/LibCrafting.sol";
+import "../../diamond/libraries/LibDiamond.sol";
+import "../../diamond/security/DiamondReentrancyGuard.sol";
+
+contract CraftingFacet is ERC1155Holder, ERC721Holder, DiamondReentrancyGuard {
+    function addRecipe(Recipe calldata recipe) public {
+        LibCrafting.CraftingStorage storage cs = LibCrafting.craftingStorage();
+        cs.numRecipes++;
+        cs.recipes[cs.numRecipes] = recipe;
+    }
+
+    function getRecipe(uint256 recipeId) public view returns (Recipe memory) {
+        return LibCrafting.craftingStorage().recipes[recipeId];
+    }
+
+    function numRecipes() public view returns (uint256) {
+        return LibCrafting.craftingStorage().numRecipes;
+    }
+
+    function craft(uint256 recipeId) public diamondNonReentrant {
+        Recipe memory recipe = LibCrafting.craftingStorage().recipes[recipeId];
+        for (uint256 i = 0; i < recipe.inputs.length; i++) {
+            if (recipe.inputs[i].tokenType == LibCrafting.ERC20_TOKEN_TYPE) {
+                ERC20Burnable token = ERC20Burnable(
+                    recipe.inputs[i].tokenAddress
+                );
+                if (
+                    recipe.inputs[i].tokenAction ==
+                    LibCrafting.INPUT_TOKEN_ACTION_TRANSFER
+                ) {
+                    token.transferFrom(
```

Comment:
We should start reading the output of `transferFrom` - it is a `bool` which is `true` if the transfer was successful.

It's possible for ERC20 token to not revert, but return `false` iff transfer was unsuccessful. For that case, we should do a check if the transfer was unsuccessful, based on return value, and revert ourselves.

Reply:
resolved